include/cocaine/rpc/asio/header.hpp

```diff
@@ -0,0 +1,545 @@
+    empty() const;
+
+    static constexpr size_t max_data_capacity = 4096;
+    static constexpr size_t http2_header_overhead = 32;
+    //32 bytes overhead per record and 2 bytes for nil-nil header
+    static constexpr size_t max_header_capacity = max_data_capacity/(http2_header_overhead + 2);
+
+private:
+    inline
+    void
+    pop();
+
+    inline
+    size_t
+    find(const std::function<bool(const header_t&)> comp);
+
+    // Header storage. Implemented as circular buffer
+    std::array<header_t, max_header_capacity> headers;
+    size_t header_lower_bound;
+    size_t header_upper_bound;
+
+    // Header data storage. Stores all data which headers can reference.
+    // Implemented as a sort of circular buffer.
+    // We multiply by 2 as data can be padded and we don't want to move it in memory.
+    // 2 multiplier guarantee that we can add new value to the end or beginning without data overlap.
+    std::array<char, max_data_capacity*2> header_data;
+    size_t data_lower_bound;
+    size_t data_lower_bound_end;
+    size_t data_upper_bound;
+    size_t capacity;
+};
+
+// Determines size of integer encoded via http2 binary protocol.
+inline
+size_t
+http2_integer_size(size_t sz, size_t bit_offset);
+
+
+// Encodes integer to dest via http2 binary protocol.
+// Note dest should be at least 10 bytes long.
+inline
+size_t
+http2_integer_encode(char* dest, size_t source, size_t bit_offset, char prefix);
+
+// --------------------------------------------------
+// Implementation part. Header-only to be used in CNF
+// --------------------------------------------------
+
+header_t::header_t(const header::data_t& _name, const header::data_t& _value) noexcept :
+    name(_name),
+    value(_value)
+{
+}
+
+header_t::header_t(const char* _name, size_t name_sz, const char* _value, size_t value_sz) noexcept :
+    name({_name, name_sz}),
+    value({_value, value_sz})
+{}
+
+header::data_t
+header_t::get_name() const {
+    return name;
+}
+
+header::data_t
+header_t::get_value() const {
+    return value;
+}
+
+size_t
+header_t::http2_size() const {
+    // 1 refer to string literals which has size with 1-bit padding.
+    // See https://tools.ietf.org/html/draft-ietf-httpbis-header-compression-12#section-5.2
+    return name.size + http2_integer_size(name.size, 1) + value.size + http2_integer_size(value.size, 1);
+}
+
+header_table_t::header_table_t() :
+    header_lower_bound(0),
+    header_upper_bound(0),
+    data_lower_bound(0),
+    data_upper_bound(0),
+    capacity(max_data_capacity)
+{}
+
+size_t
+header_table_t::size() const {
+    if(data_upper_bound > data_lower_bound) {
+        return header_static_table_t::size + data_upper_bound - data_lower_bound;
+    }
+    else {
+        return header_static_table_t::size + data_lower_bound_end - data_lower_bound + data_upper_bound;
+    }
+}
+
+bool
+header_table_t::empty() const {
+    return data_lower_bound == data_upper_bound;
+}
+
+void
+header_table_t::push(header_t& result) {
```

Comment:
Document, decompose or do something with this method. I don't understand how it works.